hnn_core/extracellular.py

```diff
@@ -0,0 +1,587 @@
+            provided.
+        voltage_offset : float | None (optional)
+            Amount to offset traces by on the voltage-axis. Useful for plotting
+            laminar arrays.
+        voltage_scalebar : float | None (optional)
+            Height, in units of uV, of a scale bar to plot in the top-left
+            corner of the plot.
+        contact_labels : list (optional)
+            Labels associated with the contacts to plot. Passed as-is to
+            :meth:`~matplotlib.axes.Axes.set_yticklabels`.
+        show : bool
+            If True, show the figure
+
+        Returns
+        -------
+        fig : instance of plt.fig
+            The matplotlib figure handle into which time series were plotted.
+        """
+        from .viz import plot_extracellular
+
+        if trial_no is None:
+            plot_data = self.get_data()
+        elif isinstance(trial_no, (list, tuple, int, slice)):
+            plot_data = self.get_data()[trial_no, ]
+        else:
+            raise ValueError(f'unkown trial number type, got {trial_no}')
+
+        if isinstance(contact_no, (list, tuple, int, slice)):
+            plot_data = plot_data[:, contact_no, ]
+        elif contact_no is not None:
+            raise ValueError(f'unkown contact number type, got {contact_no}')
+
+        for trial_data in plot_data:
+            fig = plot_extracellular(
+                self.times, trial_data, tmin=tmin, tmax=tmax, ax=ax,
+                decim=decim, color=color,
+                voltage_offset=voltage_offset,
+                voltage_scalebar=voltage_scalebar,
+                contact_labels=contact_labels,
+                show=show)
+        return fig
+
+
+class ExtracellularArrayBuilder(object):
+    """The ExtracellularArrayBuilder class
+
+    Parameters
+    ----------
+    array : ExtracellularArray object
+        The instance of :class:`hnn_core.extracellular.ExtracellularArray` to
+        build in NEURON-Python
+    """
+    def __init__(self, array):
+        self.array = array
+        self.n_contacts = array.n_contacts
+        self._nrn_imem_ptrvec = None
+        self._nrn_imem_vec = None
+        self._nrn_r_transfer = None
+        self._nrn_times = None
+        self._nrn_voltages = None
+
+    def _build(self, cvode=None):
+        """Assemble NEURON objects for calculating extracellular potentials.
+
+        The handler is set up to maintain a vector of membrane currents at at
+        every inner segment of every section of every cell on each CVODE
+        integration step. In addition, it records a time vector of sample
+        times.
+
+        Parameters
+        ----------
+        cvode : instance of h.CVode
+            Multi order variable time step integration method.
+        """
+        secs_on_rank = h.allsec()  # get all h.Sections known to this MPI rank
+        # filter list to include pyramidal cells ('Pyr') only
+        secs_on_rank = [s for s in secs_on_rank if 'Pyr' in s.name()]
```

Comment:
Let's keep this in mind before merging #357. Since our hnn-netpyne meetings, I've been trying to see where some of the implicit assumptions about the structure of our network may cause problems for expanding the model.

In this case, there would be an undocumented requirement that cells have 'Pyr' in the name.

Reply:
Also I may have missed this in discussion, but are we sure we want to exclude the inhibitory cells from the LFP!?

See [Local field potentials primarily reflect inhibitory neuron activity in human and monkey cortex](https://www.nature.com/articles/srep40211)

Reply:
You're absolutely right, it is a drastic choice to only allow pyramidal cells to contribute to the LFP! I've added some logic to `_build` in b274a6f. It's not exposed in the API, but could serve as a reminder to us to return to?

FYI you can test the effect by changing `network_builder.py:L495` to `nrn_arr._build(cvode=_CVODE, include_celltypes='all')`. Doesn't seem to have much.

There's a lot about these LFP that we don't understand! I think it will prove hard to reconcile the LFPs in our network with biologically realistic ones. I think some sort of smoothing will be needed, perhaps inspired by

Mazzoni, A. et al. [Computing the local field potential (LFP) from integrate-and-fire network models](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004584). PLOS Comput Biol 11, e1004584 (2015).

Reply:
Drive by comment without reading the papers ...

From the standpoint of the measurement, I would expect to include all cells. Maybe some of them end up canceling etc., but that's a different point. But from the standpoint of comparison, I would do only Pyramidal cells because dipoles are only inserted in the Pyramidal cells (although you could consider non-zero contributions from other cell types)

Reply:
I'm OK with setting `'all'` as the default; shall I?

Btw I just found a buglet left over from #353: the basket cell cylinders were still extending in Y instead of Z. I've fixed that here in 342b6a83 (doesn't deserve a mention).

Reply:
>  I think it will prove hard to reconcile the LFPs in our network with biologically realistic ones

but people have been doing it all along apparently. I saw a poster yesterday at the brain initiative that did exactly this. Will share the link.

This issue is yet another complexity (in addition to space, smoothing, reduced model etc) that makes the LFP validation difficult! MEG signal reflects Pyramidal neuron activity, LFP reflects inhibitory ... how do you reconcile the two for validation

Reply:
here: https://journals.plos.org/ploscompbiol/article?rev=1&id=10.1371/journal.pcbi.1008386#pcbi-1008386-g007

Btw, we should look more closely at BMTK ... they seem to be doing similar stuff as Netpyne and with (reasonable!) documentation and API

Reply:
`all` as default sounds good to me. I think this is just adding more reasons to do more basic investigations with LFP's in our model to understand precisely where these comparisons start to break down.

Also the PR is shaping up really nicely! If there's no huge rush to merge the last thing I'd want to look over is the LFP calculation code to make sure I understand it w.r.t. to the cited source.

Reply:
I've made the change in ef4e84a

Take your time! No huge rush on my part.